src/ParallelAlgorithms/LinearSolver/Gmres/ElementActions.hpp

```diff
@@ -36,15 +39,40 @@ namespace LinearSolver::gmres::detail {
 template <typename Metavariables, typename FieldsTag, typename OptionsGroup>
 struct ResidualMonitor;
 template <typename FieldsTag, typename OptionsGroup, bool Preconditioned,
-          typename Label>
+          typename Label, typename ArraySectionIdTag>
+struct PrepareStep;
+template <typename FieldsTag, typename OptionsGroup, bool Preconditioned,
+          typename Label, typename ArraySectionIdTag>
 struct NormalizeOperandAndUpdateField;
 }  // namespace LinearSolver::gmres::detail
 /// \endcond
 
 namespace LinearSolver::gmres::detail {
 
+// The linear solver may run over a section of the array parallel component,
+// i.e. over a subset of the elements. This function retrieves the section
+// that the element belongs to, or `Parallel::no_section()` if the solve is not
+// restricted to a section.
+template <typename ParallelComponent, typename ArraySectionIdTag,
+          typename DbTagsList>
+auto& get_section([[maybe_unused]] const gsl::not_null<db::DataBox<DbTagsList>*>
+                      box) noexcept {
+  if constexpr (std::is_same_v<ArraySectionIdTag, void>) {
+    return Parallel::no_section();
+  } else {
+    auto& section = db::get_mutable_reference<
```

Comment:
are you getting a reference to a pointer here or why is `section` being dereferenced below. It looks like this is a `auto * const`? If not, a comment would be useful. Even better, write out the concrete type. I think the return type can also be a concrete type

Reply:
it's a reference to an `optional`. I made the type explicit. I left the return type as `auto&` for now because it's fairly easy to see what it is now.